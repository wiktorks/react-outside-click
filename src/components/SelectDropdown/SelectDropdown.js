import React, { useState } from 'react';
import styles from './SelectDropdown.module.css';

const SelectDropdown = React.forwardRef((props, ref) => {
    const [isDisplayed, toggleVisibility] = useState(false);

    const handleDisplay = (event) => {
        console.log(event.target);
        if(!props.waitingOnClick) {
            toggleVisibility(!isDisplayed);
        }
    }

    let dropdownStyle = null;

    if(isDisplayed) {
        dropdownStyle = styles.visible;
    }

    return (
        <div className={styles['select-dropdown']} onClick={handleDisplay} ref={ref}>
            <div className={styles.title}>
                <h3>Select Dropdown</h3>
            </div>
            <p className={dropdownStyle}>{props.children}</p>
        </div>
    );
});

export default SelectDropdown;