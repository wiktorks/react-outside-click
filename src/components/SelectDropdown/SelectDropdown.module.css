.select-dropdown {
    display: inline-block;
    width: 100%;
    border: 2px solid black;
    background-color: #fff;
    position: relative;
    margin: 20px 0;
}

.select-dropdown .title {
    width: 100%;
    text-align: center;
    background-color: #B90EE4;
    color: #fff;
    padding: 20px 0;
    height: 30px;
}

.select-dropdown h3 {
    margin: 0;
}

.select-dropdown p {
    margin: 0;
    max-height: 0;
    overflow: hidden;
    width: 100%;
    top: 70px;
    right: 0;
    left: 0;
    position: absolute;
    background-color: #fff;
    border: 1px solid black;
    transition: max-height .28s;
}

.select-dropdown p.visible {
    max-height: 500px;
    
    overflow: visible;
    z-index: inherit;
}