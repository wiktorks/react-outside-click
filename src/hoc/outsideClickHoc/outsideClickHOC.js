import React, { Component } from 'react';
import style from './style.module.css';

const outsideClickHOC = (Dropdown) => class extends Component {
    constructor(props) {
        super(props);
        this.dropdownRef = React.createRef(null)
        this.state = {
            waitingOnClickOutside: false,
            wrapperClass: null,
            coverClass: style.hidden
        }
        this.preventClick = false;
    }

    onStartListeningClickOutside = () => {
        if (!this.preventClick) {
            if (!this.state.waitingOnClickOutside) {
                this.setState({
                    waitingOnClickOutside: true,
                    wrapperClass: style.wrapper,
                    coverClass: style.cover
                });
            }
        } else {
            this.preventClick = false;
        }
    }

    onClickOutside = () => {
        this.preventClick = true;
        this.setState({
            waitingOnClickOutside: false,
            wrapperClass: null,
            coverClass: style.hidden
        }, () => this.dropdownRef.current.click());
    }

    render() {
        return (
            <>
                <div
                    className={`${this.state.coverClass} cover-div`}
                    onClick={this.onClickOutside}
                ></div>
                <div
                    className={`${this.state.wrapperClass} container`}
                    style={{ display: "inline-block", width: '30%' }}
                    onClick={this.onStartListeningClickOutside}
                >
                    <Dropdown
                        ref={this.dropdownRef}
                        waitingOnClick={this.state.waitingOnClickOutside}
                        {...this.props}
                    >{this.props.children}</Dropdown>
                </div>
            </>
        );
    }
}

export default outsideClickHOC;


// Stwórz HOCa outsideClickHOC, który bedzie działać np z customowymi
// componentami typu dropdown lub select: 

// który będzie działać wg poniższych wytycznych:
// - będzie do owrapowanego komponentu dodawać propsy oraz działanie:

// działanie 1:
// -- prop - onStartListeningClickOutside 
// --- jeśli ta funkcja zostanie wywołana dany komponent zyskuje z-index najwyższy na stronie
// --- pod komponentem pojawia się półprzeźroczyste ciemne tło, w które można kliknąć zamykając komponent

// działanie 2:
// -- prop - waitingOnClickOutside
// --- zwraca true/false jeśli została wywołana metoda onStartListeningClickOutside

// działanie 3:
// -- prop onClickOutside
// --- jesli ta funkcja zostanie wywołana komponent wraca na swoje miejsce z odpowiednim z-index
// --- znika półprzeźroczyste szare tło spod komponentu

// - mentor 
// - czs na refactor
// - czas na pisanie testów
