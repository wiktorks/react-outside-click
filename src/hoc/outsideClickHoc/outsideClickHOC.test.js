
import React from "react";
import { shallow, mount } from "enzyme";
import outsideClickHOC from "./outsideClickHOC";
import styles from './style.module.css';
import SelectDropdown from "../../components/SelectDropdown/SelectDropdown";
import { configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

configure({adapter: new Adapter()});

describe("outsideClickHOC tests", () => {
    const HOC = outsideClickHOC(SelectDropdown);

    it('Should render properly', () => {
        const testHOC = shallow(<HOC />);
        expect(testHOC).toMatchSnapshot();
    });

    it('Creates background after click', () => {
        const testHOC = mount(<HOC />);
        testHOC.find('.container').simulate('click');
        expect(testHOC.find('.cover-div')).toHaveProperty("position", "fixed");
        testHOC.unmount();
    });

    it('Checks if component is hidden after clicking on background', () => {
        const testHOC = mount(<HOC />);
        testHOC.find('.container').simulate('click');
        testHOC.find('.cover-div').simulate('click');
        expect(testHOC.find('.container')).toHaveProperty("position", "static");
        testHOC.unmount();
    });
});