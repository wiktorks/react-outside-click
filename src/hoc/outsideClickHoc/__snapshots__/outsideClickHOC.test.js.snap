// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`outsideClickHOC tests Should render properly 1`] = `ShallowWrapper {}`;
